File: 3381-maximum-subarray-sum-with-length-divisible-by-k/3381-maximum-subarray-sum-with-length-divisible-by-k.py
class Solution:
    def maxSubarraySum(self, nums: List[int], k: int) -> int:
        INF = 10**30
        minPrefix = [INF] * k
        minPrefix[0] = 0

        prefix = 0
        answer = -10**30

        for i, x in enumerate(nums):
            prefix += x
            mod = (i + 1) % k

            if minPrefix[mod] != INF:
                answer = max(answer, prefix - minPrefix[mod])

            minPrefix[mod] = min(minPrefix[mod], prefix)

        return answer

File: 3381-maximum-subarray-sum-with-length-divisible-by-k/3381-maximum-subarray-sum-with-length-divisible-by-k.java
class Solution {
    public long maxSubarraySum(int[] nums, int k) {
        long[] minPrefix = new long[k];
        final long INF = Long.MAX_VALUE;

        for (int i = 0; i < k; i++) minPrefix[i] = INF;
        minPrefix[0] = 0;

        long prefix = 0;
        long answer = Long.MIN_VALUE;

        for (int i = 0; i < nums.length; i++) {
            prefix += nums[i];
            int mod = (i + 1) % k;

            if (minPrefix[mod] != INF) {
                answer = Math.max(answer, prefix - minPrefix[mod]);
            }

            minPrefix[mod] = Math.min(minPrefix[mod], prefix);
        }

        return answer;
    }
}